Examples: Add GPU Progressive Lightmapping Example

zalo opened the pull request:
Related issue: Fixed #14048

**Description**

This PR adds a light-weight example for generating silky smooth, physically-correct lightmaps over time by jittering directional lights + shadows (1st bounce only).

The lightmapping accumulation step for this scene takes between 0.2ms and 0.4ms per frame on my machine (GTX 1080).

While the setup currently generates physically accurate soft shadows and ambient occlusion, the technique could generalize to multibounce global illumination/radiosity with the right kind of depth-peeling "light sources".

Check out the live deployment here:
https://zalo.github.io/three.js/examples/?q=shadowmap#webgl_shadowmap_progressive

![image](https://user-images.githubusercontent.com/174475/110451479-15cbad80-8079-11eb-9cbf-382a93ef51c0.png)

The [potpack module](https://mapbox.github.io/potpack/) is a (very) small utility for packing multiple mesh's UVs into a unified texture atlas for the lightmap.

zalo committed:
A promising start...

But the blending is backwards!

zalo committed:
Beginning to Believe...

zalo committed:
Pre-Switching Over...

zalo committed:
Attempt to port the progressive shadowmapping to two objects...

zalo committed:
Only update shadows periodically...

zalo committed:
Why isn't this accumulating?!

zalo committed:
Excise the Effect Composer

Turned out to be a lead crutch!

zalo committed:
Get everything working, trim unused scripts

zalo committed:
Blur the edges of the Lightmap

zalo committed:
Publish Page

zalo committed:
Add progressive surface to examples?

zalo committed:
Try HalfFloats for iOS

zalo committed:
Begin hacking in iOS Support...

zalo committed:
Revert "Begin hacking in iOS Support..."

This reverts commit f98dc3975f0123845d45682a9334c7863722566a.

zalo committed:
Turn the Jitter back up

zalo committed:
Add Transform Gizmo and AO

zalo committed:
Switch AO to a Uniform Hemisphere Distribution

zalo committed:
Brighten, Enable Dithering, Switch back to HalfFloats

zalo committed:
Switch back to full floats since iOS doesn't work anyway

zalo committed:
Separate Files, Document Surface Mapping

zalo committed:
Codify Alternate Scene Universe

scene.attach() (not scene.add()) was the secret!

This is still slower than it needs to be, extraneous renders...

zalo committed:
Remove Extraneous Import

zalo committed:
Switch to Explicit Lightmapping Mode

zalo committed:
Get Lightmap Packing Working

zalo committed:
Add padding to remove bleeding

zalo committed:
Optimize Rendering; Add Debug Visualization

zalo committed:
Add back aesthetic bug

zalo committed:
Limit Debug Warnings

zalo committed:
Fix Debug Display

zalo committed:
Fix Debug Display 2

zalo committed:
Doobify and Prepare for PR

zalo committed:
Build GH Pages

Mugen87 commented:
> examples/models/obj/ShadowmappableMesh.obj 

Do you mind using a `glb` file instead?

zalo committed:
Accumulate Frames to Pass Test

Mugen87 commented:
The E2E test for `webgl_shadowmap_progressive` is still failing. Maybe you have to regenerate the screenshot.

For testing: https://raw.githack.com/zalo/three.js/progressive-lightmapping/examples/webgl_shadowmap_progressive.html

Mugen87 commented on examples/jsm/misc/ProgressiveLightMap.js:
```diff
@@ -0,0 +1,315 @@
+ * @param {WebGLRenderer} renderer A WebGL Rendering Context
+ * @param {number} res The side-long dimension of you total lightmap
+ */
+var ProgressiveLightMap = function ( renderer, res = 1024 ) {
```
Do you mind writing this code as a ES6 class?

zalo replied:
Interesting; I'm still learning when it comes to Modern Javascript Syntax, so hopefully I did that right...

Mugen87 replied:
It looks good! 👍

zalo committed:
Update the Screenshot via node test

zalo committed:
Switch mesh to a .glb, class to ES6

zalo committed:
Remove another unnecessary semicolon

zalo commented:
Happy to add additional stylization, tests, API changes, or aspects to the scene (if desired).  Don’t hesitate to put work on me 👍

Also happy to see the retweet 😁 
https://twitter.com/threejs_org/status/1369421471643271172

marcofugaro commented:
@zalo does the `ShadowmappableMesh.glb` has to be created in any particular way from blender or it can be just a regular model?

zalo committed:
Fix WebGL1 Support and Warnings

zalo commented:
@supereggbert I _think_ this was due to a WebGL1 compatibility breaking issue I just fixed in my latest commit.   But just to be safe, I added Android to the list of `userAgents` that trigger Half precision floating point textures.  The difference in quality is almost negligible, but there is more support for it on older phones.

@marcofugaro _Technically_ the only requirement is that the object is UV mapped.  However, you'll get oddly colored lines at the texture seams if two faces share an edge in the map (bilinear interpolation edge bleed), so adding some padding is ideal.  I created this model with [automatic UVs from my own program](https://github.com/zalo/CascadeStudio/pull/83), so I'm not sure if Blender has any special needs...

@Mugen87 Thank you for the suggestion to force a WebGL1 Renderer!  I'd been using `texture` instead of `texture2D` 🙃 
Correcting this fixes support on iOS (though, I don't have any Android phones to test with).  I also fixed the warnings you reported.

supereggbert commented:
> (though, I don't have any Android phones to test with)

Working here now

zalo commented:
Slightly off-topic: I'm testing out radiosity using [Approximate Radiosity by Stochastic Depth Buffering](http://xevius.org/papers/approxradpaper2010.pdf) (which is technically the multi-bounce extension of this PR's technique (that also doesn't require depth peeling)).

Early results are here:
https://raw.githack.com/zalo/three.js/progressive-radiosity/examples/?q=progress#webgl_shadowmap_progressive

It's not really appropriate for this PR or class.  It's like triple the complexity+compute, converges way more slowly, and it works best with emissive meshes rather than directional lights... all for dubious additional visual quality.  That said, I thought you guys might enjoy it.  :-)

mrdoob commented:
> That said, I thought you guys might enjoy it. :-)

Yes! 😍

mrdoob commented:
Thanks!
